### ISSUE TITLE
beam search problem

### ISSUE THREAD

**Issue by 3015218116** [brandontrabucco/detailed_captioning#3]
I read the original paper and the code you implemented, and found that the beam search stage was used in the train stage while the original text was inference stage. Could you explain the inference or the trick?

Training and Inference We perform a two-stage training strategy to train such caption generation models as proposed by Rennie et al. (2017). In the ﬁrst stage, we train the model to optimize the standard cross-entropy loss. In the second stage, we perform policy gradient with REINFORCE to optimize the non-differentiable reward function (CIDEr score in our case). **During inference, we apply beam search (beam size=2) to decode the caption**

**Comment by brandontrabucco** [comment]
Hello,

To answer your question, the implementation of the file `detailed_captioning/layers/image_captioner.py` assigns a value to a variable named `use_beam_search` on [line 33](https://github.com/brandontrabucco/detailed_captioning/blob/ae70bc91ae97e8e6033dec6d16082fce05a070f7/detailed_captioning/layers/image_captioner.py). During training, the ground truth caption is used to provide inputs to the RNN. This is called teacher forcing, and is the _first state_ of training mentioned in the original paper. When the ground truth is provided, the variable `use_beam_search` on [line 33](https://github.com/brandontrabucco/detailed_captioning/blob/ae70bc91ae97e8e6033dec6d16082fce05a070f7/detailed_captioning/layers/image_captioner.py) is false, and beam search is not used. If the ground truith is not provided (such as during inference) then `use_beam_search` is true, and beam search is used. Note than when `use_beam_search` is true, the output of the method ImageCaptioner.__call__ has an extra dimension of _beam_size_ in position 1.

As an aside, this repository does not implement the second stage of training you referenced, optimizing the CIDEr score using REINFORCE. However, one could implement this with our codebase with little change to the underlying models. 

Thanks,
Brandon

**Comment by 3015218116** [comment]
Thanks a lot ！
